### PR TITLE
[FLINK-13012][hive] Handle default partition name of Hive table

### PR DESCRIPTION
**Pull request by lirui-apache**
## What is the purpose of the change

To handle default partition name when reading/writing Hive tables.

## Brief change log

  - Handle default partition name.
  - Added a new test case for default partition name.

## Verifying this change

Comes with new test case.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

### CODE REVIEW

**Comment by lirui-apache** [comment]
The newly added test is disabled for now. I'll update this PR once #9039 is in.

**Comment by lirui-apache** [comment]
cc @xuefuz @bowenli86 @zjuwangg 
This PR handles default partition name as well as adding e2e test case. Let me know if you think we should split it into separate PRs.
And I changed hive-exec dependency to compile-scope in hive connector. It's because provided deps are not transitive, and therefore anything depending on hive connector needs to declare its own dependency on hive-exec and keep in sync with hive connector. I suppose we don't need to worry about licenses as long as we don't bundle hive in our jars, is that correct?

File: flink-connectors/flink-connector-hive/pom.xml [hunk 1]
```diff
@@ -492,9 +491,9 @@ under the License.
 						<configuration>
 							<shadeTestJar>false</shadeTestJar>
 							<artifactSet>
-								<includes>
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/pom.xml]: upon looking at the mvn shade plugin, I think we probably don't even need it anymmore, and can remove this whole section. Can you double check?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableOutputFormat.java [hunk 2]
```diff
@@ -312,7 +317,11 @@ public void writeRecord(Row record) throws IOException {
 				for (int i = dynamicPartitionOffset; i < record.getArity(); i++) {
 					// TODO: seems Hive also just calls toString(), need further investigation to confirm
 					// TODO: validate partition value
-					String partVal = record.getField(i).toString();
+					Object field = record.getField(i);
+					String partVal = field != null ? field.toString() : null;
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableOutputFormat.java]: nit: can we use full name as `partitionValue`?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableOutputFormat.java [hunk 3]
```diff
@@ -312,7 +317,11 @@ public void writeRecord(Row record) throws IOException {
 				for (int i = dynamicPartitionOffset; i < record.getArity(); i++) {
 					// TODO: seems Hive also just calls toString(), need further investigation to confirm
 					// TODO: validate partition value
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableOutputFormat.java]: do we still need the TODOs?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java [hunk 4]
```diff
@@ -160,7 +162,14 @@ private void initAllPartitions() {
 						String partitionValue = partition.getValues().get(i);
 						partitionSpec.put(partitionColName, partitionValue);
 						DataType type = catalogTable.getSchema().getFieldDataType(partitionColName).get();
-						Object partitionObject = restorePartitionValueFromFromType(partitionValue, type);
+						Object partitionObject;
+						if (defaultPartitionName.equals(partitionValue)) {
+							LogicalTypeRoot typeRoot = type.getLogicalType().getTypeRoot();
+							// while this is inline with Hive, seems it should be null for string columns as well?
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java]: can we not have a question left as comment? :)

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java [hunk 5]
```diff
@@ -160,7 +162,14 @@ private void initAllPartitions() {
...
+							partitionObject = typeRoot == LogicalTypeRoot.CHAR || typeRoot == LogicalTypeRoot.VARCHAR ? defaultPartitionName : null;
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java]: Wondering why partitionObject will be null if the type root is not char/varchar?

File: flink-end-to-end-tests/flink-connector-hive-test/pom.xml [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-hadoop-compatibility_${scala.binary.version}</artifactId>
+			<version>${project.version}</version>
+			<scope>test</scope>
+		</dependency>
+
+	</dependencies>
+
+	<!-- ONLY run the tests when explicitly told to do so  -->
```
Comment (bowenli86) [inline flink-end-to-end-tests/flink-connector-hive-test/pom.xml]: why do we need to do so?

File: flink-end-to-end-tests/flink-connector-hive-test/pom.xml [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+		<dependency>
+			<groupId>org.apache.hive.hcatalog</groupId>
+			<artifactId>hive-webhcat-java-client</artifactId>
+			<version>${hive.version}</version>
+			<scope>test</scope>
+		</dependency>
+
+		<!--  We have 0.9.10 in dependency management but hiverunner requires 0.9.8, so need to explicitly specify it here -->
+		<dependency>
```
Comment (bowenli86) [inline flink-end-to-end-tests/flink-connector-hive-test/pom.xml]: if we package a uber test-jar of flink-connector-hive, would it save us from listing all the test dependencies again here?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java (hunk 4 above)
Comment (xuefuz) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java]: Hive also has String type. What's Hive's behavior on this?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableOutputFormat.java (hunk 3 above)
Comment (xuefuz) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableOutputFormat.java]: Is this TODO comment still valid?

**Comment by xuefuz** [comment]
Re: dep scope change

I guess it's a bigger change at this point. If I understand correctly, you are trying to make the pom file for the test project simpler. The motivation seems fine, but it invalidates the purpose of us putting "provided" there in the place, right? What's the consequence to the user with this change?

File: flink-connectors/flink-connector-hive/pom.xml (hunk 1 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/pom.xml]: Yeah I think we can remove it. The shaded jar actually only contains classes from connector-hive

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableOutputFormat.java (hunk 3 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableOutputFormat.java]: Will remove them

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java (hunk 5 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java]: If the partition value equals default partition name, it means the partition value should actually be null.

File: flink-end-to-end-tests/flink-connector-hive-test/pom.xml (hunk 6 above)
Comment (lirui-apache) [reply flink-end-to-end-tests/flink-connector-hive-test/pom.xml]: It's following conventions of e2e test -- not to run any tests by default

File: flink-end-to-end-tests/flink-connector-hive-test/pom.xml (hunk 7 above)
Comment (lirui-apache) [reply flink-end-to-end-tests/flink-connector-hive-test/pom.xml]: I guess not. Test happens before package. So we won't have the uber test-jar (not sure whether we can even build one) during testing.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java (hunk 4 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSource.java]: For a string partition column, it'll be assigned the default value, e.g. `__HIVE_DEFAULT_PARTITION__`. But I thought it should simply be null, just like other kinds of partition columns.

File: flink-end-to-end-tests/flink-connector-hive-test/pom.xml (hunk 7 above)
Comment (bowenli86) [reply flink-end-to-end-tests/flink-connector-hive-test/pom.xml]: I mean building a uber test-jar of flink-connector-hive for flink-connector-hive-test, not for flink-connector-hive itself. Is it possible?

**Comment by lirui-apache** [comment]
Changed to test with table env. This PR needs to wait until we have #9181 merged.

**Comment by lirui-apache** [comment]
PR rebased.
@xuefuz @zjuwangg please take another look. Thanks.

**Comment by lirui-apache** [comment]
Thanks @xuefuz.
@wuchong could you please help review and merge this PR. Travis test has passed: https://travis-ci.com/flink-ci/flink/builds/120691298

**Comment by lirui-apache** [comment]
Thanks @wuchong for help with the merge!

**Pull request closed without merging**